src/foundry/common/utils/string-tree.d.mts

```diff
@@ -0,0 +1,81 @@
+/**
+ * A data structure representing a tree of string nodes with arbitrary object leaves.
+ */
+declare class StringTree {
+  /**
+   * The key symbol that stores the leaves of any given node.
+   * @type {symbol}
+   */
+  static get leaves(): symbol;
+
+  static #leaves: symbol;
+
+  /**
+   * The tree's root.
+   * @type {StringTreeNode}
+   */
+  #root: StringTree.StringTreeNode;
+
+  /**
+   * Create a new node.
+   * @returns   The new node
+   */
+  #createNode(): StringTree.StringTreeNode;
+
+  /**
+   * Insert an entry into the tree.
+   * @param strings  - The string parents for the entry.
+   * @param entry         - The entry to store.
+   * @returns   The node the entry was added to.
+   */
+  addLeaf(strings: string[], entry: StringTree.StringTreeNode): StringTree.StringTreeNode;
+
+  /**
+   * Traverse the tree along the given string path and return any entries reachable from the node.
+   * @param strings         - The string path to the desired node.
+   * @param options         - Additional options to configure behaviour.
+   * @param options.limit   - The maximum number of items to retrieve.
+   * @returns    The reachable entries
+   */
+  lookup(strings: string[], options?: { limit?: number }): StringTree.StringTreeNode[];
```

Comment:
I mentioned this in #2653 so look there for a further explanation if needed. Please use `InexactPartial` or add `| undefined` here.